lib/pegboard/manager/src/ctx.rs

```diff
@@ -1,23 +1,201 @@
-use anyhow::*;
+use std::{
+	collections::HashMap,
+	path::{Path, PathBuf},
+	sync::Arc,
+};
 
-use std::path::{Path, PathBuf};
-use tokio::fs;
+use anyhow::*;
+use futures_util::{
+	stream::{SplitSink, SplitStream},
+	SinkExt, StreamExt,
+};
+use indoc::indoc;
+use pegboard::protocol;
+use sqlx::{pool::PoolConnection, Sqlite, SqlitePool};
+use tokio::{fs, net::TcpStream, sync::Mutex};
+use tokio_tungstenite::{tungstenite::protocol::Message, MaybeTlsStream, WebSocketStream};
 use url::Url;
 use uuid::Uuid;
 
-use crate::utils;
+use crate::{container::Container, utils};
 
 pub struct Ctx {
 	path: PathBuf,
+	pool: SqlitePool,
+	tx: Mutex<SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>>,
+
+	containers: Mutex<HashMap<Uuid, Arc<Container>>>,
 }
 
 impl Ctx {
-	pub fn new() -> Self {
-		Ctx {
-			path: Path::new("/etc/pegboard").to_path_buf(),
+	pub fn new(
+		path: PathBuf,
+		pool: SqlitePool,
+		tx: SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>,
+	) -> Arc<Self> {
+		Arc::new(Ctx {
+			path,
+			pool,
+			tx: Mutex::new(tx),
+			containers: Mutex::new(HashMap::new()),
+		})
+	}
+
+	pub async fn sql(&self) -> std::result::Result<PoolConnection<Sqlite>, sqlx::Error> {
+		// Attempt to use an existing connection
+		if let Some(conn) = self.pool.try_acquire() {
+			std::result::Result::Ok(conn)
+		} else {
+			// Create a new connection
+			self.pool.acquire().await.map_err(Into::into)
+		}
+	}
+
+	async fn send_packet(&self, packet: protocol::ToServer) -> Result<()> {
```

Comment:
what happens on failure? how are all socket close edge cases handled/

Reply:
the manager errors then restarts and reconnects to the ws